### ISSUE TITLE
SoundEffect destructor sometimes crashes

### ISSUE THREAD

**Issue by spunky44** [MonoGame/MonoGame#5195]
The SoundEffect destructor will sometimes throw an AccessViolationException when exiting the game.

```
/// <summary>
        /// Releases unmanaged resources and performs other cleanup operations before the
        /// <see cref="Microsoft.Xna.Framework.Audio.SoundEffect"/> is reclaimed by garbage collection.
        /// </summary>
        ~SoundEffect()
        {
            Dispose(false);
        }


private void PlatformStop(bool immediate)
        {
            if (_voice != null)
            {
                if (immediate)
                {
                    _voice.Stop(0);  // <- throws AccessViolationException
                    _voice.FlushSourceBuffers();
                }
                else
                    _voice.Stop((int)PlayFlags.Tails);
            }

            _paused = false;
        }
```

This is on Windows.

**Comment by Jjagg** [comment]
Probably related to #5153, since you get this exception when you attempt to access voices when their master voice is destroyed.

**Comment by spunky44** [comment]
It happens fairly regularly, both when running in VS and when deployed.

**Comment by spunky44** [comment]
Some users have found these crashes in their Windows log. The exception doesn't result in an error dialog it seems.

**Comment by Jjagg** [comment]
Well the OS should block any attempt to read protected memory, so maybe it doesn't propagate to the application somehow... However that's a bit weird. From https://msdn.microsoft.com/en-us/library/ms164911.aspx:

    Access violations in unmanaged code that bubble up to managed code are **always** wrapped in an AccessViolationException.

Maybe finalizers behave differently or maybe it's a threading thing? I really don't know.

**Comment by RefactoredGames** [comment]
I just updated to the newest develop version of MonoGame and this error still exists.

**Comment by KonajuGames** [comment]
What are the repro steps for this? I've never seen this crash and I've just
completed a whole bunch of work with SoundEffect.

**Comment by KonajuGames** [comment]
Is there a sound playing at the time of exit? Have played and completed
sounds?

**Comment by Jjagg** [comment]
I can reproduce this by calling `SoundEffectInstance.Stop()` after `Game` is disposed. We should check if the MasterVoice is disposed in `SoundEffectInstance` methods.

**Comment by Jjagg** [comment]
@RefactoredGames How do you manage your SoundEffectInstances? When do you call Stop on them? Do you manage them in a seperate thread?

**Comment by RefactoredGames** [comment]
This error comes when I exit the game. I don't call stop on any sound effect instances when exiting.

But could this problem be thread related? The music player had crashes too and they were solved by setting locks around volume setting and so on.

**Comment by RefactoredGames** [comment]
I keep references to all playing sound effect instances, but interact with them in the main thread.

**Comment by Jjagg** [comment]
The issue is that in XAudio you have a MasteringVoice and SourceVoices. A SourceVoice is connected to a MasteringVoice. In MG when we clean up audio, we dispose of the MasteringVoice which cleans up all associated SourceVoices too, but MG does not clean up the managed SoundEffectInstances, so they do not know that their SourceVoice reference is no longer valid. When we then try to do something with a SourceVoice when it is no longer valid we get an AccessViolationException. I just opened pull request #5827, that makes SoundEffectInstance check if we disposed of the MasteringVoice before calling functions on a SourceVoice, so that should fix the issue.
